hub/core/transform/transform.py

```diff
@@ -0,0 +1,166 @@
+            For example:- transform(data_in, [fn1, fn2, fn3], ds_out, [{"a":5, "b":6}, {}, {"c":11, "s":7}]), only applies kwargs to fn1 and fn3.
+        scheduler (str): The scheduler to be used to compute the transformation.
+            Currently can be one of:-
+            threaded: Uses multithreading to perform the transform. Best applicable for I/O intensive transforms.
+            processed: Uses multiprocessing to perform the transform. Best applicable for CPU intensive transforms. Currently doesn't work with S3 or Hub cloud datasets.
+        workers (int): The number of workers to use for performing the transform. Defaults to 1.
+
+    Raises:
+        InvalidInputDataException: If ds_in passed to transform is invalid. It should support __getitem__ and __len__ operations.
+        TensorMismatchException: If one or more of the outputs generated during transform contain different tensors than the ones present in the output 'ds_out' provided to transform.
+        UnsupportedSchedulerException: If the scheduler passed is not recognized.
+        InvalidTransformException: If the output of any step in a transformation isn't dictionary or a list/tuple of dictionaries.
+    """
+    ds_out.flush()
+
+    if not hasattr(data_in, "__getitem__"):
+        raise InvalidInputDataException("__getitem__")
+    if not hasattr(data_in, "__len__"):
+        raise InvalidInputDataException("__len__")
+
+    # TODO: this check doesn't work currently. Will work once AL-1092 is merged and can be uncommented.
+    # for tensor in tensors:
+    #     assert len(ds_out[tensor]) == len(ds_out)
+
+    workers = max(workers, 1)
+
+    if scheduler == "threaded":
+        compute: ComputeProvider = ThreadProvider(workers)
+    elif scheduler == "processed":
+        compute = ProcessProvider(workers)
+    else:
+        raise UnsupportedSchedulerException(scheduler)
+
+    base_storage = remove_all_cache(ds_out.storage)
+    tensors = set(ds_out.meta.tensors)
+
+    pipeline_kwargs = pipeline_kwargs or []
+    pipeline_kwargs = pipeline_kwargs[0 : len(pipeline)]
+    pipeline_kwargs += [{}] * (len(pipeline) - len(pipeline_kwargs))
+
+    store(data_in, base_storage, tensors, compute, workers, pipeline, pipeline_kwargs)
+    ds_out.clear_cache()
+    ds_out._load_meta()
+
+
+def store_shard(transform_input: Tuple):
+    """Takes a shard of the original data and iterates through it, producing chunks."""
+    data_shard, size, storage, tensors, pipeline, pipeline_kwargs = transform_input
+
+    # storing the metas in memory to merge later
+    all_index_meta = {key: IndexMeta.create(key, MemoryProvider()) for key in tensors}
+    all_tensor_meta = {key: TensorMeta.create(key, MemoryProvider()) for key in tensors}
+
+    # separate cache for each tensor to prevent frequent flushing, 32 MB ensures only full chunks are written.
+    storage_map = {key: LRUCache(MemoryProvider(), storage, 32 * MB) for key in tensors}
```

Comment:
no change required here: but would it make sense to have a prefetch cache for reading and an LRUcache for writing?